### PR TITLE
fix to restore artifacts from .nemo for ner

### PR DESCRIPTION
**Pull request by ekmb**

### CODE REVIEW

File: nemo/collections/nlp/models/token_classification/token_classification_model.py [hunk 1]
```diff
@@ -43,16 +47,28 @@ def input_types(self) -> Optional[Dict[str, NeuralType]]:
...
+        modes = ["train", "test", "dev"]
+        if not os.path.exists(cfg.data_desc_pickle):
+            # process the data and extract label_ids and class weights if the model is not restored from .nemo file
+            self.data_desc = TokenClassificationDataDesc(
+                data_dir=self.data_dir, modes=modes, pad_label=cfg.dataset.pad_label
+            )
+            if not self.data_desc.data_found:
+                raise ValueError(f'No {modes} data found at {self.data_dir}.')
+        else:
+            self.data_desc = pickle.load(open(cfg.data_desc_pickle, "rb"))
```
Comment (titu1994) [inline nemo/collections/nlp/models/token_classification/token_classification_model.py:69]: Use register artifact function to properly return a correct path rather than do manually checks

File: nemo/collections/nlp/data/token_classification/token_classification_descriptor.py [hunk 2]
```diff
@@ -34,16 +34,25 @@ def __init__(self, data_dir: str, modes: List[str] = ['train', 'test', 'dev'], p
...
+            )
 
+        self.data_dir = data_dir
+        # when the model is restored from .nemo file, data might be missing, and data_desc values should be skipped
+        self.data_found = False
         unique_labels = set()
+
         for mode in modes:
             all_labels = []
             label_file = os.path.join(data_dir, 'labels_' + mode + '.txt')
```
Comment (titu1994) [inline nemo/collections/nlp/data/token_classification/token_classification_descriptor.py:50]: Why not use register artifact here to serialize the label file too and use return value to test for if it exists or not ? Then data fiu d flag becomes unnecessary

File: nemo/collections/nlp/modules/common/bert_module.py [hunk 3]
```diff
@@ -50,7 +51,12 @@ def save_to(self, save_path: str):
 
     def restore_weights(self, restore_path: str):
         """Restores module/model's weights"""
-        logging.info(f"restore from {restore_path}")
+        logging.info(f"Restoring weights from {restore_path}")
+
+        if not os.path.exists(restore_path):
+            logging.warning(f'Path {restore_path} not found')
+            return
```
Comment (titu1994) [inline nemo/collections/nlp/modules/common/bert_module.py:58]: This part is not called by core automatically, but explicitly called by user. It should fail fast if they give wrong paths

File: nemo/core/classes/modelPT.py [hunk 4]
```diff
@@ -190,14 +190,29 @@ def restore_from(cls, restore_path: str):
...
                 instance.load_state_dict(torch.load(model_weights))
 
+                logging.info(f'Model {cls.__name__} was successfully restored from {restore_path}.')
             finally:
                 os.chdir(cwd)
 
         return instance
 
+    @classmethod
+    def update_config_with_specific_artifacts(cls, config: OmegaConf, artifacts_dir: str = None) -> OmegaConf:
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:204]: No need for artigscts_dir to be None by default, it is assured provided as tmpdir if it gets called

File: nemo/core/classes/modelPT.py [hunk 5]
```diff
@@ -190,14 +190,29 @@ def restore_from(cls, restore_path: str):
                 model_weights = path.join(tmpdir, _MODEL_WEIGHTS)
                 conf = OmegaConf.load(config_yaml)
                 OmegaConf.set_struct(conf, True)
+                cls.update_config_with_specific_artifacts(config=conf, artifacts_dir=tmpdir)
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:193]: This needs conf = cls.update... to match return signature of the method.

File: nemo/collections/nlp/data/token_classification/token_classification_descriptor.py (hunk 2 above)
Comment (ekmb) [reply nemo/collections/nlp/data/token_classification/token_classification_descriptor.py:50]: the label file could be huge

File: nemo/collections/nlp/models/token_classification/token_classification_model.py (hunk 1 above)
Comment (ekmb) [reply nemo/collections/nlp/models/token_classification/token_classification_model.py:69]: In the config I provide only the name of the data_desc_pickle, I create the self.data_desc during data preprocessing step that is done during the model init, so I have to store self.data_desc first

File: nemo/collections/nlp/models/token_classification/token_classification_model.py [hunk 6]
```diff
@@ -158,6 +174,16 @@ def validation_epoch_end(self, outputs):
     def setup_training_data(self, train_data_config: Optional[DictConfig]):
         self._train_dl = self._setup_dataloader_from_config(cfg=train_data_config, prefix='train')
 
+        if not torch.distributed.is_initialized() or torch.distributed.get_rank() == 0:
```
[the remark this answers is not in the document]
Comment (ekmb) [reply nemo/collections/nlp/models/token_classification/token_classification_model.py:177]: I don't see why, the second conditional will be checked only if the first one fails

File: nemo/collections/nlp/models/token_classification/token_classification_model.py [hunk 7]
```diff
@@ -158,6 +174,16 @@ def validation_epoch_end(self, outputs):
...
+            self.register_artifact('label_ids.csv', self.data_desc.label_ids_filename)
+
+            if self._cfg.data_desc_pickle is None:
+                self._cfg.data_desc_pickle = 'data_desc.p'
+
+            data_desc_file = os.path.join('/tmp', self._cfg.data_desc_pickle)
+            pickle.dump(self.data_desc, open(data_desc_file, 'wb'))
+            self.register_artifact(config_path='model.data_desc_pickle', src=data_desc_file)
```
[the remark this answers is not in the document]
Comment (ekmb) [reply nemo/collections/nlp/models/token_classification/token_classification_model.py:185]: I don't have the file yet

File: nemo/collections/nlp/modules/common/bert_module.py (hunk 3 above)
Comment (ekmb) [reply nemo/collections/nlp/modules/common/bert_module.py:58]: When the model is trained from a provided bert_checkpoint path, it will store this to the config. And then during restore_from(), the file could be missing (this is to fix yesterday's error, we discussed in slack). I agree, the fast fail would be better, but then another hack to make restoration form .nemo file is needed. I can remove the missing checkpoint path from the config wit the update_config..() call, but every nlp model should do the same then. @titu1994 do you think it's better to that?

File: nemo/core/classes/modelPT.py (hunk 5 above)
Comment (ekmb) [reply nemo/core/classes/modelPT.py]: fixed

File: nemo/core/classes/modelPT.py (hunk 4 above)
Comment (ekmb) [reply nemo/core/classes/modelPT.py]: fixed

File: nemo/collections/nlp/models/token_classification/token_classification_model.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (titu1994) [reply nemo/collections/nlp/models/token_classification/token_classification_model.py:177]: Not for an or clause. PyTorch is lazy only for and

File: nemo/collections/nlp/models/token_classification/token_classification_model.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (titu1994) [reply nemo/collections/nlp/models/token_classification/token_classification_model.py:185]: Interesting. Alright that's reasonable then.

File: nemo/collections/nlp/data/token_classification/token_classification_descriptor.py (hunk 2 above)
Comment (titu1994) [reply nemo/collections/nlp/data/token_classification/token_classification_descriptor.py:50]: Ah makes sense

File: nemo/collections/nlp/models/token_classification/token_classification_model.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/nlp/models/token_classification/token_classification_model.py:69]: Ok thats another issue then, the setup_* methods can be called multiple times (either if it supports multiple data loaders which I think this script not support thankfully, and during .restore_from()). I see that the file being saved is not marked with unique name, so basically if it gets called multiple time, previous values will get overwritten. Is it possible to refactor this outside of the setup method, or to provide a unique name inside it?

File: nemo/collections/nlp/modules/common/bert_module.py (hunk 3 above)
Comment (titu1994) [reply nemo/collections/nlp/modules/common/bert_module.py:58]: Ok this is for restoration code path, makes sense. Theres two ways to approach this
1) "clean but heavy way" - add these checkpoints as artifacts and make the .nemo model massive in size (especially for larger BERT/ variants). This is kinda bad cause the actual NER model is just a fraction of the size.
2) "workaround way" - as shown here, it shifts burden to individual models and logs missing filepaths. If every single NLP model is going to do this, I would suggest having a baseclass for NLP (like how ASR has the unique ASRModel which is what all other models subclass from. In this NLP base class, you could do this work around.

I say this NLP Baseclass is a better approach, in the future user who wishes to build models using NeMo for NLP should have the least friction. They should not have to consider such manifest workarounds for a good user experience of saving and restoring .nemo files. 

The NLP refactor can be done in a future PR, for now this workaround is ok.

File: nemo/collections/nlp/models/token_classification/token_classification_model.py (hunk 1 above)
Comment (ekmb) [reply nemo/collections/nlp/models/token_classification/token_classification_model.py:69]: I'm saving this file inside setup training data only, why would it be called multiple times? Didn't you disable setup_data calls for restore_from()?

Comment (titu1994) [reply nemo/collections/nlp/models/token_classification/token_classification_model.py:69]: Setup disable calls are attempted by the restore from step (since it's in the constructor). It will fail on other machines and print a log message since paths won't match, and so I guess if won't save file either. That's fine then.
